compiler/optimizer/VPHandlers.cpp

```diff
@@ -8998,169 +8998,67 @@ static void removeConditionalBranch(OMR::ValuePropagation *vp, TR::Node *node, T
-            newConstraint = NULL;
-            }
-         else
-            {
-            TR::VPObjectLocation *loc = NULL;
-            if (castIsClassObject)
-               {
-               // the object could have no type to begin with,
-               // if so, do *not* create the specialClass constraint
-               //
-               if (objectRefConstraint->getClassType())
-                  newType = TR::VPResolvedClass::create(vp, (TR_OpaqueClassBlock *)VP_SPECIALKLASS);
-               else
-                  newType = NULL;
-               }
-            else if (castIsInterfaceImplementedByClass)
-               {
-               // dont want to propagate the interface implemented by Class
-               // as the type if the object is a classobject
-               if (objectRefConstraint->isClassObject() == TR_yes)
-                  newType = NULL;
-               }
-
-            // inject a classObject property on the constraint
-            if ((objectRefConstraint && (objectRefConstraint->isClassObject() == TR_yes)) ||
-                  castIsClassObject)
-               loc = TR::VPObjectLocation::create(vp, TR::VPObjectLocation::ClassObject);
-
-            // final construction
-            newConstraint = TR::VPClass::create(vp, newType, presence, NULL, NULL, loc);
-            }
-         }
-      }
-   else
-      {
-      if (vp->trace())
-         traceMsg(vp->comp(), "ObjectRef has no constraint, so applying cast class properties...\n");
-
-      TR::VPClassType *newType = NULL;
-      bool castIsClassObject = false;
-      if (classConstraint->getClassType()->asResolvedClass())
-         {
-         // check if cast class is a java/lang/Class type
-         //
-         TR_OpaqueClassBlock *klass = classConstraint->getClassType()->getClass();
-         TR_OpaqueClassBlock *jlClass = vp->fe()->getClassClassPointer(klass);
-         if (jlClass)
-            {
-            applyType = true;
-            if (klass == jlClass)
-               {
-               newType = NULL;
-               castIsClassObject = true;
-               }
-            else
-               newType = TR::VPResolvedClass::create(vp, classConstraint->getClass());
-            }
-         else
-            applyType = false;
-         }
-      else
-         {
-         newType = classConstraint->getClassType();
-         applyType = true;
-         }
-      if (applyType)
-         {
-         TR::VPObjectLocation *loc = NULL;
-         if (castIsClassObject)
-            loc = TR::VPObjectLocation::create(vp, TR::VPObjectLocation::JavaLangClassObject);
+passingTypeTestObjectConstraint(
+   OMR::ValuePropagation *vp,
+   TR::VPConstraint *classConstraint,
+   bool testingForFixedType)
+   {
+   TR_ASSERT_FATAL(
+      classConstraint->isClassObject() == TR_yes,
+      "expected a instanceof classConstraint to be a 'ClassObject'");
+
+   TR::VPClassType *type = classConstraint->getClassType();
+   TR_ASSERT_FATAL(
+      type != NULL,
```

Comment:
Class.isInstance gets mapped to an instanceof IL opcode in the IL gen in some cases, and in the most general case, this could generate an instanceof with the second child not being a constant class. In this case, is it possible that we don't know the type ? i.e. we know it's a ClassObject but we don't know of what type.

Reply:
The only caller is `constrainIfcmpeqne()`. At both call sites it passes `instanceofConstraint`, and only when `instanceofConstraint` has been set, in which case it always has a `getClassType()`. We could make `passingTypeTestObjectConstraint()` more general, so that it handles completely unknown types, but it would only be able to tell us that the object is non-null, and the generality would go unused without further changes elsewhere.

Reply:
Okay so the callers of the routine currently always pass in a class type. Thanks for the clarification.